Framework/Algorithms/src/SampleCorrections/MCInteractionVolume.cpp

```diff
@@ -56,6 +61,49 @@ const Geometry::BoundingBox &MCInteractionVolume::getBoundingBox() const {
   return m_sample->getBoundingBox();
 }
 
+/**
+ * Generate point randomly across one of the components of the environment
+ * including the sample itself in the selection. The method first selects
+ * a random component and then selects a random point within that component
+ * using Object::generatePointObject
+ * @param rng A reference to a PseudoRandomNumberGenerator where
+ * nextValue should return a flat random number between 0.0 & 1.0
+ * @return The generated point
+ */
+Kernel::V3D MCInteractionVolume::generatePoint(
+    Kernel::PseudoRandomNumberGenerator &rng) const {
+  for (size_t i = 0; i < m_maxScatterAttempts; i++) {
+    Kernel::V3D point;
+    int componentIndex;
+    if (m_env) {
+      componentIndex = rng.nextInt(0, static_cast<int>(m_env->nelements())) - 1;
+    } else {
+      componentIndex = -1;
+    }
+
+    bool pointGenerated;
+    if (componentIndex == -1) {
+      pointGenerated =
+          m_sample->generatePointInObject(rng, m_activeRegion, 1, point);
+    } else {
+      pointGenerated =
+          m_env->getComponent(componentIndex)
+              .generatePointInObject(rng, m_activeRegion, 1, point);
+    }
+    if (pointGenerated) {
+      if (componentIndex == -1) {
+        m_sampleScatterPoints++;
+      } else {
+        m_envScatterPoints[componentIndex]++;
+      }
+      return point;
+    }
+  }
+  throw std::runtime_error("MCInteractionVolume::generatePoint() - Unable to "
```

Comment:
More of a higher level question, is there no way to specify bounds to the RNG generator so that it generates fixed points within a region of interest, rather than discarding points until we find one?

Reply:
I think the discarding of points is a positive thing here because it supports the "fair" sampling I'm trying to achieve. It allows the code to randomly generate points within the active region (which is a box) and then ask the question "is this point inside each part of the environment?". This trial\error process results in the bigger components getting more scatter points generated in them.

To answer your question specifically - for simple shapes (sphere, cylinder) etc you can generate points that are both within the region of interest and inside the shape in a single attempts without discarding points. This is essentially what the fallback method in CSObject::generatePointInObject is doing that we have discussed elsewhere on this PR. But having a method that always finds a point first time isn't that helpful in achieving fair sampling between the environment components.

For non-simple objects (meshes) I don't think there is a way of generating points first time that are both in the region of interest and inside the object

Reply:
Ahh okay, I was fundamentally misreading this as we generate points for a single component rather than a collection of components within a bounded region.

This is a really good explanation thanks 👍